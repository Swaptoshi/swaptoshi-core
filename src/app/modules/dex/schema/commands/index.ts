export * from './burn_command';
export * from './collect_command';
export * from './collect_treasury_command';
export * from './create_pool_command';
export * from './decrease_liquidity_command';
export * from './exact_input_command';
export * from './exact_input_single_command';
export * from './exact_output_command';
export * from './exact_output_single_command';
export * from './increase_liquidity_command';
export * from './mint_command';
export * from './treasurify_command';
