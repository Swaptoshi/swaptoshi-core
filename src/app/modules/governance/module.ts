/* eslint-disable */
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import { BaseModule, BlockAfterExecuteContext, BlockExecuteContext, ModuleInitArgs, ModuleMetadata, TokenMethod } from 'klayr-sdk';
import { GovernanceEndpoint } from './endpoint';
import { GovernanceMethod } from './method';
import { GovernanceInternalMethod } from './internal_method';
import { TreasuryMintEvent } from './events/treasury_mint';
import { TreasuryBlockRewardTaxEvent } from './events/treasury_block_reward_tex';
import { ConfigUpdatedEvent } from './events/config_updated';
import { GovernableConfigRegistry } from './registry';
import { GovernanceGovernableConfig } from './config';

export class GovernanceModule extends BaseModule {
	public endpoint = new GovernanceEndpoint(this.stores, this.offchainStores);
	public method = new GovernanceMethod(this.stores, this.events);
	public commands = [];

	private _config = new GovernanceGovernableConfig(this.name, 0);
	private _governableConfig = new GovernableConfigRegistry();
	private _internalMethod = new GovernanceInternalMethod(this.stores, this.events);

	public constructor() {
		super();
		this.stores.register(GovernanceGovernableConfig, this._config);

		this.events.register(TreasuryMintEvent, new TreasuryMintEvent(this.name));
		this.events.register(ConfigUpdatedEvent, new ConfigUpdatedEvent(this.name));
		this.events.register(TreasuryBlockRewardTaxEvent, new TreasuryBlockRewardTaxEvent(this.name));
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [],
			assets: [],
		};
	}

	public async init(args: ModuleInitArgs): Promise<void> {
		this.method.init(this._governableConfig);
		this.method.registerGovernableConfig(args, this.name, this._config);
	}

	public addDependencies(token: TokenMethod) {
		this._internalMethod.addDependencies(token, this._governableConfig);
	}

	public async afterTransactionsExecute(context: BlockAfterExecuteContext): Promise<void> {
		await this._internalMethod.addTreasuryReward(context);
	}

	public async beforeTransactionsExecute(context: BlockExecuteContext): Promise<void> {
		await this._internalMethod.initializeGovernableConfig(context);
	}
}
