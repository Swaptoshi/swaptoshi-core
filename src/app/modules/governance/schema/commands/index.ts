export * from './create_proposal_command';
export * from './vote_command';
export * from './boost_vote_command';
export * from './delegate_vote_command';
export * from './revoke_delegated_vote_command';
export * from './set_proposal_attributes';
