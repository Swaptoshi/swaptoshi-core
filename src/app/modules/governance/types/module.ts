import { FeeMethod, TokenMethod } from 'klayr-sdk';

export interface GovernanceModuleDependencies {
	tokenMethod: TokenMethod;
	feeMethod: FeeMethod;
}

export type VoteScoreArray = { voteScore: bigint; boostingHeight: number }[];

export type VoteScoreOrArray = bigint | VoteScoreArray;
