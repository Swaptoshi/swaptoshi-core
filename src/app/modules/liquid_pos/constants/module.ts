export const POS_MODULE_NAME = 'pos';
export const POS_STAKE_COMMAND_NAME = 'stake';
