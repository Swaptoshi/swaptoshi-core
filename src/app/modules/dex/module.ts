/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import {
	BaseInteroperableModule,
	BlockExecuteContext,
	FeeMethod,
	MainchainInteroperabilityMethod,
	ModuleInitArgs,
	ModuleMetadata,
	NFTMethod,
	SidechainInteroperabilityMethod,
	TokenMethod,
	TransactionExecuteContext,
	TransactionVerifyContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { BurnCommand } from './commands/burn_command';
import { CollectCommand } from './commands/collect_command';
import { CreatePoolCommand } from './commands/create_pool_command';
import { DecreaseLiquidityCommand } from './commands/decrease_liquidity_command';
import { ExactInputCommand } from './commands/exact_input_command';
import { ExactInputSingleCommand } from './commands/exact_input_single_command';
import { ExactOutputCommand } from './commands/exact_output_command';
import { ExactOutputSingleCommand } from './commands/exact_output_single_command';
import { IncreaseLiquidityCommand } from './commands/increase_liquidity_command';
import { MintCommand } from './commands/mint_command';
import { TreasurifyCommand } from './commands/treasurify_command';
import { DexEndpoint } from './endpoint';
import { BurnEvent } from './events/burn';
import { CollectEvent } from './events/collect';
import { CollectPositionEvent } from './events/collect_position';
import { CollectProtocolEvent } from './events/collect_protocol';
import { DecreaseLiquidityEvent } from './events/decrease_liquidity';
import { FlashEvent } from './events/flash';
import { IncreaseLiquidityEvent } from './events/increase_liquidity';
import { IncreaseObservationCardinalityNextEvent } from './events/increase_observation_cardinality_next';
import { MintEvent } from './events/mint';
import { PoolCreatedEvent } from './events/pool_created';
import { PoolInitializedEvent } from './events/pool_initialized';
import { SwapEvent } from './events/swap';
import { TokenURICreatedEvent } from './events/tokenuri_created';
import { TokenURIDestroyedEvent } from './events/tokenuri_destroyed';
import { DexMethod } from './method';
import {
	getPoolEndpointRequestSchema,
	getPoolEndpointResponseSchema,
} from './schema/endpoint/get_pool';
import {
	getPositionEndpointRequestSchema,
	getPositionEndpointResponseSchema,
} from './schema/endpoint/get_position';
import {
	getTokenURIEndpointRequestSchema,
	getTokenURIEndpointResponseSchema,
} from './schema/endpoint/get_token_uri';
import {
	observeEndpointRequestSchema,
	observeEndpointResponseSchema,
} from './schema/endpoint/observe';
import {
	quoteExactInputEndpointRequestSchema,
	quoteExactInputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input';
import {
	quoteExactInputSingleEndpointRequestSchema,
	quoteExactInputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input_single';
import {
	quoteExactOutputEndpointRequestSchema,
	quoteExactOutputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output';
import {
	quoteExactOutputSingleEndpointRequestSchema,
	quoteExactOutputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output_single';
import { ObservationStore } from './stores/observation';
import { PoolStore } from './stores/pool';
import { PositionInfoStore } from './stores/position_info';
import { PositionManagerStore } from './stores/position_manager';
import { TickBitmapStore } from './stores/tick_bitmap';
import { TickInfoStore } from './stores/tick_info';
import { TokenSymbolStore } from './stores/token_symbol';
import { DexModuleConfig } from './types';
import { verifyMinimumFee, verifySwapByTransfer, executeSwapByTransfer } from './hooks';
import { defaultConfig } from './constants';
import {
	getMetadataEndpointRequestSchema,
	getMetadataEndpointResponseSchema,
} from './schema/endpoint/get_metadata';
import { verifyFeeConversion } from './hooks/verifyFeeConversion';
import { executeFeeConversion } from './hooks/executeFeeConversion';
import { TreasurifyEvent } from './events/treasurify';
import {
	getPoolAddressFromCollectionIdEndpointRequestSchema,
	getPoolAddressFromCollectionIdEndpointResponseSchema,
} from './schema/endpoint/get_pool_address_from_collection_id';
import {
	quotePriceEndpointRequestSchema,
	quotePriceEndpointResponseSchema,
} from './schema/endpoint/quote_price';
import { TokenRegisteredEvent } from './events/token_registered';
import { SupportedTokenStore } from './stores/supported_token';
import { verifyValidTransfer } from './hooks/verifyValidTransfer';
import { DexInteroperableMethod } from './cc_method';

export class DexModule extends BaseInteroperableModule {
	public _config: DexModuleConfig | undefined;
	public _feeMethod: FeeMethod | undefined;
	public _tokenMethod: TokenMethod | undefined;
	public _dexInteroperableMethod = new DexInteroperableMethod(this.stores, this.events);

	public crossChainCommand = [];
	public crossChainMethod = this._dexInteroperableMethod;

	public endpoint = new DexEndpoint(this.stores, this.offchainStores);
	public method = new DexMethod(this.stores, this.events);
	public commands = [
		new CreatePoolCommand(this.stores, this.events),
		new MintCommand(this.stores, this.events),
		new IncreaseLiquidityCommand(this.stores, this.events),
		new DecreaseLiquidityCommand(this.stores, this.events),
		new CollectCommand(this.stores, this.events),
		new BurnCommand(this.stores, this.events),
		new ExactInputCommand(this.stores, this.events),
		new ExactInputSingleCommand(this.stores, this.events),
		new ExactOutputCommand(this.stores, this.events),
		new ExactOutputSingleCommand(this.stores, this.events),
		new TreasurifyCommand(this.stores, this.events),
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(PoolStore, new PoolStore(this.name, 0, this.stores, this.events));
		this.stores.register(
			PositionManagerStore,
			new PositionManagerStore(this.name, 1, this.stores, this.events),
		);

		this.stores.register(ObservationStore, new ObservationStore(this.name, 2));
		this.stores.register(PositionInfoStore, new PositionInfoStore(this.name, 3));
		this.stores.register(TickBitmapStore, new TickBitmapStore(this.name, 4));
		this.stores.register(TickInfoStore, new TickInfoStore(this.name, 5));
		this.stores.register(TokenSymbolStore, new TokenSymbolStore(this.name, 6, this.events));
		this.stores.register(SupportedTokenStore, new SupportedTokenStore(this.name, 7));

		this.events.register(BurnEvent, new BurnEvent(this.name));
		this.events.register(CollectPositionEvent, new CollectPositionEvent(this.name));
		this.events.register(CollectProtocolEvent, new CollectProtocolEvent(this.name));
		this.events.register(CollectEvent, new CollectEvent(this.name));
		this.events.register(DecreaseLiquidityEvent, new DecreaseLiquidityEvent(this.name));
		this.events.register(FlashEvent, new FlashEvent(this.name));
		this.events.register(IncreaseLiquidityEvent, new IncreaseLiquidityEvent(this.name));
		this.events.register(
			IncreaseObservationCardinalityNextEvent,
			new IncreaseObservationCardinalityNextEvent(this.name),
		);
		this.events.register(MintEvent, new MintEvent(this.name));
		this.events.register(PoolCreatedEvent, new PoolCreatedEvent(this.name));
		this.events.register(PoolInitializedEvent, new PoolInitializedEvent(this.name));
		this.events.register(SwapEvent, new SwapEvent(this.name));
		this.events.register(TokenURICreatedEvent, new TokenURICreatedEvent(this.name));
		this.events.register(TokenURIDestroyedEvent, new TokenURIDestroyedEvent(this.name));
		this.events.register(TreasurifyEvent, new TreasurifyEvent(this.name));
		this.events.register(TokenRegisteredEvent, new TokenRegisteredEvent(this.name));
	}

	public addDependencies(
		tokenMethod: TokenMethod,
		nftMethod: NFTMethod,
		feeMethod: FeeMethod,
		interoperabilityMethod: SidechainInteroperabilityMethod | MainchainInteroperabilityMethod,
	) {
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportManagerStore = this.stores.get(SupportedTokenStore);

		poolStore.addDependencies(tokenMethod);
		positionManagerStore.addDependencies(tokenMethod, nftMethod);
		supportManagerStore.addDependencies(tokenMethod);

		this._feeMethod = feeMethod;
		this._tokenMethod = tokenMethod;
		this._dexInteroperableMethod.addDependencies(interoperabilityMethod, tokenMethod, nftMethod);
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [
				{
					name: this.endpoint.getPool.name,
					request: getPoolEndpointRequestSchema,
					response: getPoolEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPosition.name,
					request: getPositionEndpointRequestSchema,
					response: getPositionEndpointResponseSchema,
				},
				{
					name: this.endpoint.getTokenURI.name,
					request: getTokenURIEndpointRequestSchema,
					response: getTokenURIEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPoolAddressFromCollectionId.name,
					request: getPoolAddressFromCollectionIdEndpointRequestSchema,
					response: getPoolAddressFromCollectionIdEndpointResponseSchema,
				},
				{
					name: this.endpoint.getMetadata.name,
					request: getMetadataEndpointRequestSchema,
					response: getMetadataEndpointResponseSchema,
				},
				{
					name: this.endpoint.observe.name,
					request: observeEndpointRequestSchema,
					response: observeEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInput.name,
					request: quoteExactInputEndpointRequestSchema,
					response: quoteExactInputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInputSingle.name,
					request: quoteExactInputSingleEndpointRequestSchema,
					response: quoteExactInputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutput.name,
					request: quoteExactOutputEndpointRequestSchema,
					response: quoteExactOutputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutputSingle.name,
					request: quoteExactOutputSingleEndpointRequestSchema,
					response: quoteExactOutputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quotePrice.name,
					request: quotePriceEndpointRequestSchema,
					response: quotePriceEndpointResponseSchema,
				},
			],
			assets: [],
		};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async init(_args: ModuleInitArgs): Promise<void> {
		this._config = { ...defaultConfig, ..._args.moduleConfig };
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportedTokenStore = this.stores.get(SupportedTokenStore);

		poolStore.init(this._config);
		positionManagerStore.init(Buffer.from(_args.genesisConfig.chainID, 'hex'));
		supportedTokenStore.init(this._config);
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
		try {
			await verifyMinimumFee(_context, this._config!);
			await verifyValidTransfer.bind(this)(_context);
			await verifySwapByTransfer.bind(this)(_context);
			await verifyFeeConversion.bind(this)(_context);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await verifyMinimumFee(_context, this._config!);
		await verifyValidTransfer.bind(this)(_context);
		await verifySwapByTransfer.bind(this)(_context);
		await executeFeeConversion.bind(this)(_context);
	}

	public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await executeSwapByTransfer.bind(this)(_context);
	}

	public async beforeTransactionsExecute(context: BlockExecuteContext): Promise<void> {
		await this.stores.get(SupportedTokenStore).apply(context);
	}
}
