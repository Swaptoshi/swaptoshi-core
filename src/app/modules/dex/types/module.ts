/* eslint-disable import/no-cycle */
import { FeeConversionMethod } from '../../fee_conversion';
import { GovernanceMethod } from '../../governance';
import { SidechainInteroperabilityMethod, MainchainInteroperabilityMethod, FeeMethod, TokenMethod, NFTMethod } from './utils';

export interface DexModuleDependencies {
	tokenMethod: TokenMethod;
	nftMethod: NFTMethod;
	feeMethod: FeeMethod;
	interoperabilityMethod: SidechainInteroperabilityMethod | MainchainInteroperabilityMethod;
	feeConversionMethod?: FeeConversionMethod;
	governanceMethod?: GovernanceMethod;
}
