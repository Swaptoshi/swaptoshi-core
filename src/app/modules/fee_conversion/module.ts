/* eslint-disable */
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import { BaseModule, ModuleInitArgs, ModuleMetadata, TransactionExecuteContext, TransactionVerifyContext, VerificationResult, VerifyStatus } from 'klayr-sdk';
import { FeeConversionEndpoint } from './endpoint';
import { FeeConversionMethod } from './method';
import { FeeConversionMethodRegistry } from './registry';
import {
	dryRunTransactionEndpointRequestSchema,
	dryRunTransactionEndpointResponseSchema,
	getConfigEndpointRequestSchema,
	getConfigEndpointResponseSchema,
	getRegisteredHandlersEndpointRequestSchema,
	getRegisteredHandlersEndpointResponseSchema,
} from './schema';
import { FeeConvertedEvent } from './events/fee_converted';
import { InternalFeeConversionMethod } from './internal_method';
import { FeeConversionGovernableConfig } from './config';
import { GovernanceMethod } from '../governance';
import { FeeConversionModuleDependencies } from './types';

export class FeeConversionModule extends BaseModule {
	public endpoint = new FeeConversionEndpoint(this.stores, this.offchainStores);
	public method = new FeeConversionMethod(this.stores, this.events);
	public commands = [];

	private _config = new FeeConversionGovernableConfig(this.name, 0);
	private _handler = new FeeConversionMethodRegistry();
	private _internalMethod = new InternalFeeConversionMethod(this.stores, this.events);
	private _governanceMethod: GovernanceMethod | undefined;

	public constructor() {
		super();
		this.stores.register(FeeConversionGovernableConfig, this._config);

		this.events.register(FeeConvertedEvent, new FeeConvertedEvent(this.name));
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [
				{
					name: this.endpoint.getConfig.name,
					request: getConfigEndpointRequestSchema,
					response: getConfigEndpointResponseSchema,
				},
				{
					name: this.endpoint.getRegisteredHandlers.name,
					request: getRegisteredHandlersEndpointRequestSchema,
					response: getRegisteredHandlersEndpointResponseSchema,
				},
				{
					name: this.endpoint.dryRunTransaction.name,
					request: dryRunTransactionEndpointRequestSchema,
					response: dryRunTransactionEndpointResponseSchema,
				},
			],
			assets: [],
		};
	}

	// Lifecycle hooks
	public async init(_args: ModuleInitArgs): Promise<void> {
		this.method.init(this._handler);
		this.endpoint.init(this._handler);
		await this._internalMethod.init(this._handler);
		await this._internalMethod.checkDependencies();

		if (this._governanceMethod) {
			this._governanceMethod.registerGovernableConfig(_args, this.name, this._config);
		}
	}

	public addDependencies(dependencies: FeeConversionModuleDependencies) {
		this._governanceMethod = dependencies.governanceMethod;
		this._internalMethod.addDependencies(dependencies.feeMethod, dependencies.tokenMethod, dependencies.dexMethod);
		this.endpoint.addDependencies(this._internalMethod);
	}

	public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
		try {
			await this._internalMethod.verify(_context);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await this._internalMethod.execute(_context);
	}
}
