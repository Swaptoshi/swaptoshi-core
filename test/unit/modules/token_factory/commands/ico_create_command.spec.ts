import { IcoCreateCommand } from '../../../../../src/app/modules/token_factory/commands/ico_create_command';
import { TokenFactoryModule } from '../../../../../src/app/modules/token_factory/module';

describe('IcoCreateCommand', () => {
	let module: TokenFactoryModule;
	let command: IcoCreateCommand;

	beforeEach(() => {
		module = new TokenFactoryModule();
		command = new IcoCreateCommand(module.stores, module.events);
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toBe('icoCreate');
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
			it.todo('should throw errors for invalid schema');
			it.todo('should be ok for valid schema');
		});
	});

	describe('execute', () => {
		describe('valid cases', () => {
			it.todo('should update the state store');
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
