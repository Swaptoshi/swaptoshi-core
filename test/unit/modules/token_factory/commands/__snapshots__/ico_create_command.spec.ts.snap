// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IcoCreateCommand constructor should have valid schema 1`] = `
{
  "$id": "/tokenFactory/command/icoCreate",
  "properties": {
    "amount": {
      "dataType": "uint64",
      "fieldNumber": 4,
    },
    "price": {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "providerAddress": {
      "dataType": "bytes",
      "fieldNumber": 5,
      "format": "klayr32",
    },
    "tokenIn": {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
    "tokenOut": {
      "dataType": "bytes",
      "fieldNumber": 2,
    },
  },
  "type": "object",
}
`;
